### ISSUE TITLE
GPU Passtrough problem

### ISSUE THREAD

**Issue by mrceresa** [lxc/lxd#6030]
# Required information

 * Distribution: Fedora
 * Distribution version: 29
 * The output of "lxc info" or if that fails:
```
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  driver: lxc
  driver_version: 3.2.1
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "false"
  kernel_version: 5.1.11-200.fc29.x86_64
  lxc_features:
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    seccomp_notify: "true"
  project: default
  server: lxd
  server_clustered: false
  server_pid: 25243
  server_version: "3.15"
  storage: btrfs
  storage_version: "4.4"
```

# Issue description
I installed lxd using snap on Fedora 29.
I'm trying to pass my gpu in this way: (elaborated from [here](https://asciinema.org/a/174076))
* lxc launch -t m3.large images:ubuntu/18.10/amd64 ub
* lxc config device add ub gpu gpu id=0
* lxc config set ub nvidia.runtime true

However, I get two errors:
(1) from config device:
Error: Failed to add mount for device: Failed to run: /snap/lxd/11405/bin/lxd forkmount lxc-mount ub /var/snap/lxd/common/lxd/containers /var/snap/lxd/common/lxd/logs/ub/lxc.conf /var/snap/lxd/common/lxd/devices/ub/unix.gpu.dev-dri-card0 /dev/dri/card0 none 4096: 

The permissions on my gpu are:
$ ll /dev/dri/card0
`crw-rw----+ 1 root video 226, 0 Jul  1 15:12 /dev/dri/card0`

And I'm able to stat it read only:
$ python3
```
Python 3.7.3 (default, May 11 2019, 00:45:16) 
[GCC 8.3.1 20190223 (Red Hat 8.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> fd = os.open("/dev/dri/card0", os.O_RDONLY)
>>> os.fstat(fd)
os.stat_result(st_mode=8624, st_ino=20175, st_dev=6, st_nlink=1, st_uid=0, st_gid=39, st_size=0, st_atime=1561986760, st_mtime=1561986760, st_ctime=1563784141)
```
(2) from nvidia.runtime
Error: Initialize LXC: The NVIDIA LXC hook couldn't be found
In this case, following the info [here](https://discuss.linuxcontainers.org/t/problems-trying-to-use-nvidia-runtime-with-snap-lxd-3-0-0/1750/4) I was able to enter the snap context, and effectively, I have an error:
$ /var/lib/snapd/hostfs/usr/bin/nvidia-container-cli
`/var/lib/snapd/hostfs/usr/bin/nvidia-container-cli: error while loading shared libraries: libnvidia-container.so.1: cannot open shared object file: No such file or directory`

However, outside of the snap everything works:
$ /usr/bin/nvidia-container-cli info
```
 NVRM version:   430.26
 CUDA version:   10.2
 
 Device Index:   0
 Device Minor:   0
 Model:          TITAN V
 Brand:          TITAN
 GPU UUID:       GPU-7cf83cf6-60bc-42b6-d06f-09ea4a40df50
 Bus Location:   00000000:01:00.0
 Architecture:   7.0
```
 

And the library is correctly linked
$ ldd /usr/bin/nvidia-container-cli 
`	libnvidia-container.so.1 => /usr/bin/../lib64/libnvidia-container.so.1 (0x00007ff3671fc000)`

Thanks for any help with this! I'm a bit lost on how to debug it further :)

Mario

# Information to attach

The device exists and it is visible from the host:
$ nvidia-smi 
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 430.26       Driver Version: 430.26       CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  TITAN V             Off  | 00000000:01:00.0  On |                  N/A |
| 32%   47C    P2    37W / 250W |    750MiB / 12063MiB |      3%      Default |
+-------------------------------+----------------------+----------------------+

I add the logs (under lxd --debug) of the the two errors.
[error1_DEVICE_ADD_GPU.txt](https://github.com/lxc/lxd/files/3452399/error1_DEVICE_ADD_GPU.txt)
[error2_NVIDIARUNTIME.txt](https://github.com/lxc/lxd/files/3452401/error2_NVIDIARUNTIME.txt)

**Comment by stgraber** [comment]
There is a reasonable chance that this is failing due to a kernel bug in the 5.1 kernel.

Can you try a simple:

    lxc config device add ub test unix-char source=/dev/kmsg

If that fails with the same error, then it's indeed a kernel issue.

@brauner found that cross-mntns mounts were broken and this got resolved in 5.2 I believe. I don't know whether your particular distribution has taken this fix though.

**Comment by mrceresa** [comment]
Thanks @stgraber! It is confirmed: with 5.1 it was not working, upgrading to last version solved the problem both for test and for the real gpu.
Now it is working :)
